pkg/adaptor/etcd/reader.go

```diff
@@ -0,0 +1,107 @@
+package etcd
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/compose/transporter/pkg/client"
+	"github.com/compose/transporter/pkg/log"
+	"github.com/compose/transporter/pkg/message"
+	"github.com/compose/transporter/pkg/message/data"
+	"github.com/compose/transporter/pkg/message/ops"
+
+	eclient "github.com/coreos/etcd/client"
+)
+
+var (
+	_ client.Reader = &Reader{}
+)
+
+// Reader implements client.Reader for the use of iterating an etcd directory.
+type Reader struct {
+	rootKey string
+}
+
+type nodeMsg struct {
+	node *eclient.Node
+	path string
+}
+
+func newReader(rootKey string) client.Reader {
+	return &Reader{rootKey}
+}
+
+func (r *Reader) Read(filterFn client.NsFilterFunc) client.MessageChanFunc {
+	return func(s client.Session, done chan struct{}) (chan message.Msg, error) {
+		out := make(chan message.Msg)
+		session := s.(*Session)
+		log.With("rootKey", r.rootKey).Infoln("starting Read func")
+		results := make(chan *nodeMsg)
+		go func() {
+			r.walkNode(r.rootAsPath(), session, results)
+			close(results)
+		}()
+		go func() {
+			defer close(out)
+			directoryMap := make(map[string]map[string]interface{})
+			for {
+				select {
+				case <-done:
+					return
+				case result, ok := <-results:
+					if !ok {
```

Comment:
This walking logic is quite complex but looks good. I can't say with certainty that it is correct... going to have to rely on tests.